okio/src/commonMain/kotlin/okio/RawSource.kt

```diff
@@ -72,4 +75,30 @@ interface RawSource : Closeable {
    */
   @Throws(IOException::class)
   override fun close()
+
+  /**
+   * Awaits the underlying source to fill-in the data and returns the number of bytes
+   * available, so the next `read(..., result)` will be served immediately.
+   * Returns `-1` if the underlying source is closed or exhausted.
+   *
+   * This method is optional and is allowed to return `-1` unconditionally for definitely
+   * blocking sources.
+   *
+   * ### List of open questions
+   *
+   * * Should we distinguish "exhausted/closed" from "not supported" (`-2`)?
+   *    Maybe it should even throw as it doesn't make any sense to call it on blocking sources at all.
+   * * Should we make this method optional?
+   *     * Pros -- implementing RawSource is pretty straightforward
+   *     * Cons -- easy to forget to delegate and then composability is broken
+   *     * My vote -- make it non-optional (it is optional now to reduce the patch size)
+   * * This method feels slightly off:
+   *     * We have `request()` and `require()` on `Source`
+   *     * And this single and alone `await` here
+   *     * It is perfectly understandable from an implementation standpoint,
+   *       but may be a bit weird for users.
+   */
+  suspend fun awaitAvailable(atLeastBytes: Long = -1L): Long {
```

Comment:
We can delegate this method to `read` with a zero capacity buffer by default.

We also don't need any numbers in the result: just a flag if the source is exhausted.

Reply:
But read is supposed to be non-blocking and non-suspending, could you please elaborate on how this is the reasonable default behaviour for an arbitrary source?

Also, it requires changing `read` contract which now can read as many bytes as it wants in its internal structures and give to `read`'s `buffer` only requested number of bytes

Reply:
The number of bytes in the result may (?) be important in order to know how many bytes is allowed to be read via `read` from a purely non-blocking context, in order to e.g. parse part of bytes and wait for the rest of it.

This is exactly scenario from the second commit:

* Consider a protocol wants to read length-prefixed UTF-8 string
* It reads length, it calls `awaitAvailable(length)`, but stream is malformed and returns less bytes than length

Three options here:
* Use "boolean flag", invoke "readUtf8StringStrict", fail because there is no `\n`
* Parse byte-by-byte manually and decode it, throwing where necessary. It neglects the library experience and makes authors to do the same work all over again
* Check returned number of bytes and throw a proper exception, potentially reading leftover bytes via separate byte-by-byte codepath for better error message

Reply:
Another option: use the same signature as `read`, but make it suspending.

```
suspend fun readSuspending(buffer: Buffer, byteCount: Long): Long
```

Then the buffering class, `Source`, can use this to build other primitives on top: `requestSuspending` and `requireSuspending`, and maybe even `readUtf8LineSuspending`.

The benefit is it's structurally simple. The enormous drawback is it leads us to building everything 2x, and callers now need to choose everywhere.

Reply:
It's indeed structurally much simpler, but I'm not sure if the `suspend` overhead for protocol parsing is reasonable enough to advocate for such approach as the default one